### ISSUE TITLE
Examples or documentation of inference in C++ with backends 

### ISSUE THREAD

**Issue by grisaitis** [pytorch/executorch#7291]
### 📚 The doc issue

I'm trying to make a small C++ demo of using executorch to run Llama 3.2 1B Instruct, just using C++.

Is there any documentation anywhere on, eg, how to use the backends for running inference in C++? Eg with XNNExecutor? I have no idea what the API is, let alone how to use it, unfortunately.

I downloaded, quantized, and "lowered" the model to the XNNPACK backend with:

```
export MODEL_DIR="/Users/william/.cache/huggingface/hub/models--meta-llama--Llama-3.2-1B-Instruct/snapshots/9213176726f574b556790deb65791e0c5aa438b6" 

python -m examples.models.llama2.export_llama \         
        --checkpoint ${MODEL_DIR}/original/consolidated.00.pth \
        -p ${MODEL_DIR}/original/params.json \
        -kv --use_sdpa_with_kv_cache \
        -X \
        -qmode 8da4w  \
        --group_size 128 \
        -d fp32 \
        --metadata '{"get_bos_id":128000, "get_eos_ids":[128009, 128001]}' \
        --embedding-quantize 4,32 \
        --max_seq_length 1024 \
        --output_name="llama3.2_1B_kv_sdpa_xnn_qe_4_32_context1024.pte"
```

i also built ExecuTorch for my macos M1 machine with:
```
./install_requirements.sh
./backends/apple/coreml/scripts/install_requirements.sh
./backends/apple/mps/install_requirements.sh
./build/build_apple_frameworks.sh --coreml --mps --optimized --xnnpack
./backends/xnnpack/scripts/install_requirements.sh
```

i can get the following basic C++ project to build:

```cpp
// main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>

#include <executorch/extension/llm/tokenizer/tiktoken.h>
#include <executorch/extension/module/module.h>
#include <executorch/extension/tensor/tensor.h>
#include <executorch/runtime/core/exec_aten/exec_aten.h>
#include <executorch/runtime/core/evalue.h>

using namespace executorch::extension;
using namespace executorch::extension::llm;
using executorch::aten::ScalarType;
using executorch::runtime::EValue;

std::string generate(
    Module& model,
    const std::string& prompt,
    Tiktoken& tokenizer,
    size_t max_tokens = 100) {
    
    // Encode the prompt
    auto tokens_result = tokenizer.encode(prompt, true, true);
    if (!tokens_result.ok()) {
        throw std::runtime_error("Failed to encode prompt");
    }
    std::vector<uint64_t> input_tokens = tokens_result.get();
    std::vector<uint64_t> output_tokens;

    // Initialize output accumulator
    std::string output;
    
    for (size_t i = 0; i < max_tokens; ++i) {
        // Create input tensor
        auto input_tensor = from_blob(
            input_tokens.data(),
            {1, static_cast<int>(input_tokens.size())},
            ScalarType::Long);

        // Run inference
        auto result = model.forward(input_tensor);
        if (!result.ok()) {
            throw std::runtime_error("Model inference failed");
        }

        // Get logits from output
        auto logits_tensor = result->at(0).toTensor();
        const float* logits = logits_tensor.const_data_ptr<float>();
        
        // Get the last token's logits
        const size_t vocab_size = logits_tensor.size(2);
        const float* token_logits = logits + (logits_tensor.size(1) - 1) * vocab_size;

        // Simple greedy sampling (take argmax)
        uint64_t next_token = 0;
        float max_logit = token_logits[0];
        for (size_t j = 1; j < vocab_size; ++j) {
            if (token_logits[j] > max_logit) {
                max_logit = token_logits[j];
                next_token = j;
            }
        }

        // Decode the token
        auto token_text = tokenizer.decode(
            input_tokens.empty() ? 0 : input_tokens.back(),
            next_token);
        if (!token_text.ok()) {
            throw std::runtime_error("Failed to decode token");
        }

        // Print the token
        std::cout << token_text.get();
        std::cout.flush();

        // Store the token
        output_tokens.push_back(next_token);
        
        // Update input context
        input_tokens.push_back(next_token);
    }

    std::cout << std::endl;
    return output;
}

int main() {
    try {
        // Initialize special tokens for the tokenizer
        auto special_tokens = std::make_unique<std::vector<std::string>>();
        special_tokens->push_back("<s>");    // BOS token
        special_tokens->push_back("</s>");   // EOS token
        
        // Create and load tokenizer
        Tiktoken tokenizer(std::move(special_tokens), 0, 1); // BOS index 0, EOS index 1
        auto status = tokenizer.load(TOKENIZER_PATH);
        if (status != executorch::runtime::Error::Ok) {
            std::cerr << "Failed to load tokenizer\n";
            return 1;
        }

        // Load model
        Module model(MODEL_PATH);

        // Get input from user
        std::cout << "Enter prompt: ";
        std::string prompt;
        std::getline(std::cin, prompt);

        // Generate text
        generate(model, prompt, tokenizer);
        
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
}
```

```cmake
# CMakeLists.txt

cmake_minimum_required(VERSION 3.13)
project(executorch_llm_test)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

# ExecuTorch paths
set(USER_SOURCE_DIR "/Users/william/src")
set(EXECUTORCH_SOURCE_DIR "${USER_SOURCE_DIR}/executorch")
set(EXECUTORCH_ROOT "${USER_SOURCE_DIR}/executorch/cmake-out")

# Define framework paths for macOS ARM64
set(EXECUTORCH_FRAMEWORK "${EXECUTORCH_ROOT}/executorch.xcframework/macos-arm64")
set(BACKEND_XNNPACK_FRAMEWORK "${EXECUTORCH_ROOT}/backend_xnnpack.xcframework/macos-arm64")
set(KERNELS_OPTIMIZED_FRAMEWORK "${EXECUTORCH_ROOT}/kernels_optimized.xcframework/macos-arm64")

# Set include directories
set(EXECUTORCH_INCLUDE_DIR "${EXECUTORCH_FRAMEWORK}/Headers")

# Build TikToken library
set(TIKTOKEN_SOURCES
    ${EXECUTORCH_SOURCE_DIR}/extension/llm/tokenizer/tiktoken.cpp
    ${EXECUTORCH_SOURCE_DIR}/extension/llm/tokenizer/bpe_tokenizer.cpp
)

add_library(tiktoken STATIC ${TIKTOKEN_SOURCES})

# Set include directories for tiktoken
target_include_directories(tiktoken PUBLIC  # Changed from PRIVATE to PUBLIC
    ${EXECUTORCH_SOURCE_DIR}
    ${EXECUTORCH_INCLUDE_DIR}
    ${USER_SOURCE_DIR}
)

# Find RE2 package
find_package(re2 REQUIRED)
target_link_libraries(tiktoken PUBLIC re2::re2)  # Changed from PRIVATE to PUBLIC

# Main executable
set(SOURCES
    main.cpp
)

add_executable(llm_test ${SOURCES})

# Include directories for main executable - now matches tiktoken's includes
target_include_directories(llm_test PRIVATE
    ${CMAKE_CURRENT_SOURCE_DIR}
    ${EXECUTORCH_SOURCE_DIR}
    ${EXECUTORCH_INCLUDE_DIR}
    ${USER_SOURCE_DIR}  # Added this line to match tiktoken's includes
)

# Link libraries
target_link_libraries(llm_test PRIVATE
    "${EXECUTORCH_FRAMEWORK}/libexecutorch-macos-release.a"
    "${BACKEND_XNNPACK_FRAMEWORK}/libbackend_xnnpack-macos-release.a"
    "${KERNELS_OPTIMIZED_FRAMEWORK}/libkernels_optimized-macos-release.a"
    tiktoken
)

# Define model file paths as compile definitions
target_compile_definitions(llm_test PRIVATE
    MODEL_PATH="${CMAKE_CURRENT_SOURCE_DIR}/model_source_files/llama3.2_1B_kv_sdpa_xnn_qe_4_32_context1024.pte"
    TOKENIZER_PATH="${CMAKE_CURRENT_SOURCE_DIR}/model_source_files/tokenizer.model"
)
```

it builds, but when i attempt to run it, it says:

```
➜  build git:(main) ✗ ./llm_test 
ExecuTorch PAL must be initialized before call to et_pal_current_ticks()[1]    36509 abort      ./llm_test
```

### Suggest a potential alternative/fix

give a couple simple self-contained examples like the nanogpt example below, but where the model was quantized and lowered to a particular backend.

https://pytorch.org/executorch/stable/llm/getting-started.html

**Comment by grisaitis** [comment]
Addendum: it would be great to understand how to use the `Module` interface, when and how to call `torch::executor::runtime_init();`, and also whether and how I would use `Runner` classes. 

I just wish there were more documentation how to take advantage of these APIs, aside from hunting through source files. And I don't have the cash to burn on an swarm agentic AI code analyzer to investigate for me :)

**Comment by grisaitis** [comment]
My apologies - I opened this issue due to a misunderstanding about a key aspect of backend initialization in ExecuTorch.

My misunderstanding leading to this issue: I thought that in order to run a "lowered" model (lowered to some backend) that I needed to explicitly provision or initialize a runtime for the backend. I now understand that this is done by ExecuTorch automatically - inside `Method`, I think. And this makes sense. After all, PyTorch Modules can be exported to be only partially delegated to a backend. 

For those looking for nice, small examples of running "lowered" models, I suggest searching for `*_executor_runner` in `.md`, `.cpp`, `.h`, or `target.bzl` files. E.g. `xnn_executor_runner` here: https://github.com/pytorch/executorch/blob/v0.4.0/examples/xnnpack/README.md . In the example, you can see how the `main` function running the XNNPACK model doesn't include anything in the source about backends or delegates. The executable is built, however, linked to the xnnpack_backend library built by cmake. For other backends, there are several nice examples using toy `torch.Module` modules (e.g. add two tensors or a single Conv2d op), exporting it to a backend, and running it with one of the low level API (`Program`, `Method`, `MemoryManager`, etc) or the `Module` API.
